network/src/application/interface.rs

```diff
@@ -79,80 +79,73 @@ pub trait NetworkInterface {
-        recipients: impl Iterator<Item = PeerNetworkId>,
-        protocol: ProtocolId,
-        message: TMessage,
-    ) -> Result<(), NetworkError>;
-
-    async fn send_rpc(
-        &mut self,
-        recipient: PeerNetworkId,
-        protocol: ProtocolId,
-        req_msg: TMessage,
-        timeout: Duration,
-    ) -> Result<TMessage, RpcError>;
+#[derive(Clone)]
+struct MultiNetworkSender<
+    TMessage: Message + Send,
+    Sender: ApplicationNetworkSender<TMessage> + Send,
+> {
+    senders: HashMap<NetworkId, Sender>,
+    _phantom: PhantomData<TMessage>,
 }
 
-struct MultiNetworkSender<TMessage: Message> {
-    senders: HashMap<NetworkId, NetworkSender<TMessage>>,
-}
-
-impl<TMessage: Message> MultiNetworkSender<TMessage> {
-    fn sender(&mut self, network_id: &NetworkId) -> &mut NetworkSender<TMessage> {
+impl<TMessage: Message + Send, Sender: ApplicationNetworkSender<TMessage> + Send>
+    MultiNetworkSender<TMessage, Sender>
+{
+    fn sender(&mut self, network_id: &NetworkId) -> &mut Sender {
         self.senders.get_mut(network_id).expect("Unknown NetworkId")
     }
 }
 
 #[async_trait]
-impl<TMessage: Clone + Message + Send> PeerNetworkIdSender<TMessage>
-    for MultiNetworkSender<TMessage>
+impl<TMessage: Clone + Message + Send, Sender: ApplicationNetworkSender<TMessage> + Send>
+    ApplicationPeerNetworkIdSender<TMessage> for MultiNetworkSender<TMessage, Sender>
 {
-    fn send_to(
-        &mut self,
-        recipient: PeerNetworkId,
-        protocol: ProtocolId,
-        message: TMessage,
-    ) -> Result<(), NetworkError> {
+    fn send_to(&mut self, recipient: PeerNetworkId, message: TMessage) -> Result<(), NetworkError> {
         self.sender(&recipient.network_id())
-            .send_to(recipient.peer_id(), protocol, message)
+            .send_to(recipient.peer_id(), message)
     }
 
     fn send_to_many(
         &mut self,
         recipients: impl Iterator<Item = PeerNetworkId>,
-        protocol: ProtocolId,
         message: TMessage,
     ) -> Result<(), NetworkError> {
         for (network_id, recipients) in
             &recipients.group_by(|peer_network_id| peer_network_id.network_id())
         {
             let sender = self.sender(&network_id);
             let peer_ids = recipients.map(|peer_network_id| peer_network_id.peer_id());
-            sender.send_to_many(peer_ids, protocol, message.clone())?;
+            sender.send_to_many(peer_ids, message.clone())?;
         }
         Ok(())
     }
 
     async fn send_rpc(
         &mut self,
         recipient: PeerNetworkId,
-        protocol: ProtocolId,
         req_msg: TMessage,
         timeout: Duration,
     ) -> Result<TMessage, RpcError> {
         self.sender(&recipient.network_id())
-            .send_rpc(recipient.peer_id(), protocol, req_msg, timeout)
+            .send_rpc(recipient.peer_id(), req_msg, timeout)
             .await
     }
 }
+
+#[async_trait]
+trait ApplicationPeerNetworkIdSender<TMessage: Send>: Clone {
```

Comment:
might be nice to add some documentation around a trait or within this file of how this is used, why it is used etc

Reply:
Yeah, can do